### PR TITLE
jb: java perf monitoring

### LINKED ISSUE #9521: Expose perf details for the current workspace
When users are diagnosing challenges in a workspace, it would be good to expose container-specific performance information to help the user understand what is going on with their workspace and make necessary changes. 

Ideally, this would be exposed as part of `gp cli` as this command can then be consumed via plugins in the various IDE's for IDE specific experience (e.g. the JetBrains integrated performance monitor)

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

See [internal doc](https://www.notion.so/gitpod/Java-Performance-e14d1acd4993410288b95e5687411172) for results of perf testing.

This PR adds tooling which can be used by us or end users to control and monitor resources consumptions of workspaces generally and particularly JB backends:
- adds a supervisor endpoint to fetch resources status (memory and cpu) with respect of cgroup v1
- provide internal CLI command `/.supervisor/supervisor top` to monitor resources status based on endpoint
- contributes env var for each JB product to control max heap size (XMX) via `[product]_XMX`, i.e. `INTELLIJ_XMX=4096M`
- monitor JB backends max/used memory in bytes and report it to prometheus for each workspace pod
- contributes a dev time script `node ./dev/ide/profile-workspace.js` which capture avg/max/90% percentile of memory and CPU usage as well as max/used JB backend memory for a single workspace running on VM

**Out of scope**
- https://github.com/gitpod-io/gitpod/issues/10043
- support of cgroup v2 for resources endpoint [(add a link to new issue here)]
- making it convenient to configure XMX for JB products via .gitpod.yml [[1](https://github.com/gitpod-io/gitpod/issues/8704#issuecomment-1123718534)]
- adding cgroup based stats to JB perf indicator [(add a link to new issue here)]

## Related Issue(s)

Fixes #9521 

## How to test

**Important** only one person can test at the time. So comment below that you are testing and when done remove a comment.

- Start a dev workspace [https://gitpod.io#https://github.com/gitpod-io/gitpod/tree/refs/heads/ak/java_perf](https://gitpod.io/#https://github.com/gitpod-io/gitpod/tree/refs/heads/ak/java_perf)
- Run `./dev/preview/install-k3s-kubeconfig.sh` to configure k8s context.
- Run `./dev/preview/portforward-monitoring-satellite.sh -c harvester` to port forward prometheus API endpoint.
- Run `node ./dev/ide/profile-workspace.js` to start monitoring resource usage of a workspace pod in prev envs.
- Configure prev envs with IntelliJ IDEA: https://ak-java-perf.preview.gitpod-dev.com/preferences
- Start a workspace in prev envs: https://ak-java-perf.preview.gitpod-dev.com/#INTELLIJ_XMX=4096m,JB_MEM_PROFILE=true/https://github.com/eclipse/xtext-core
- Connect with IntelliJ.
- Run `curl http://localhost:22999/_supervisor/v1/status/resources` to check resource status of supervisor.
- Use `/.supervisor/supervisor top` to see resources status of the current workspace via internal CLI
- Capture CPU, memory and heap metrics for the backend from a profile file in the dev workspace: `watch tail -n 1 /workspace/gitpod/dev/ide/perf.log`

## Release Notes

```release-note
NONE
```

## Documentation

- [x] /werft with-vm
- [ ] /werft without-vm
- [ ] /werft with-clean-slate-deployment

### CODE REVIEW

**Comment by roboquat** [comment]
@akosyakov: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/supervisor-api/status.proto
Comment (akosyakov) [inline components/supervisor-api/status.proto:76]: @csweichel Could you have a look at new status endpoint please?

In this PR it respects only cgroup v1. I will open an issue to respect cgroup v2 as a follow-up.

Comment (akosyakov) [reply components/supervisor-api/status.proto:76]: I added `top` only to `/.supervisor/supevisor`, not yet to `gp`.

https://github.com/gitpod-io/gitpod/issues/9521 mention though ideally we should have on `gp` too. Would it be alright with you to add it there?  cc @loujaybee

**Comment by mustard-mh** [comment]
I'm testing this PR 🙌

**Comment by mustard-mh** [comment]
Follow up How to test section, works well for me
|curl|top-cli|watch-log|
|-|-|-|
|curl|top-cli| watch-log|

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt
Comment (mustard-mh) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:73]: `/tmp` folder is calc as memory for our workspace. If we keet appending content, mem will keep increase, once user open workspace for a long time, it'll be a bad experience. It's more terrible than process mem increase (OOM can kill it)

Comment (akosyakov) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:73]: yeah, overwriting will be better 👍 

I actually would like to move it to prometheus reporting instead, i.e. look up it on supervisor metrics endpoint

Should I do it in this PR?

Comment (mustard-mh) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:73]: - If history is not very useful, we can replace all content
- Or make this log rotate delete?

Comment (akosyakov) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:73]: it was not useful to me, but in grafana dashboard we will be able to see history too.

Comment (mustard-mh) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:73]: If you consider this feature as normal users not reachable, we can go with it👍

**Comment by mustard-mh** [comment]
I'm testing this PR

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt
Comment (mustard-mh) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:73]: `prometheus` is good place 👍

**Comment by mustard-mh** [comment]
How to check if `xmx=4096m` works? @akosyakov

**Comment by akosyakov** [comment]
> How to check if xmx=4096m works? @akosyakov

It is a bit tricky:
- in performance indicator click ... and then select main window. It will render UI of backend in another window.
- In main window press double shift and then enable `Show Memory Usage` toggle
- Check max heap size in the status bar of this window.

**Comment by mustard-mh** [comment]
Tested and works well (we don't need expected result in how to test section?), should we explain what `/workspace/gitpod/dev/ide/perf.log` and `/tmp/jb_mem_profile.log` use for?

**Comment by akosyakov** [comment]
Do you mean a link to https://www.notion.so/gitpod/Java-Performance-e14d1acd4993410288b95e5687411172#a24a5c9f95e043edbbfb249eb086a66a?

File: components/ide/jetbrains/image/status/main.go
Comment (mustard-mh) [inline components/ide/jetbrains/image/status/main.go:271]: Should we change `2048m` to `-Xmx2048m`? It's like a magic number here if you not access `launcher.sh` file

Comment (mustard-mh) [reply components/ide/jetbrains/image/status/main.go:271]: And add this one https://github.com/gitpod-io/gitpod/blob/a4f50fa74f24c6abe1fa04a6a9e1ea6d0fee3c62/components/ide/jetbrains/image/startup.sh#L13

File: components/supervisor/pkg/supervisor/top.go
Comment (mustard-mh) [inline components/supervisor/pkg/supervisor/top.go:36]: Maybe we can use `procfs` package?

File: dev/ide/profile-workspace.js
Comment (mustard-mh) [inline dev/ide/profile-workspace.js:29]: `wsPodName` only setup once. we need to restart script next time to get correct pod name.

You can ignore it since it's not important

**Review by mustard-mh — APPROVED** [review]
/hold

Code looks good

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt
Comment (mustard-mh) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt:73]: Good for me

File: components/supervisor/pkg/supervisor/top.go
Comment (csweichel) [inline components/supervisor/pkg/supervisor/top.go:34]: We should harmonise this with the way ws-daemon does it.
Either copy over some of that code, or stick it in something like `common-go/cgroup`.

@Furisto @utam0k could you guide here how to proceed?

Comment (Furisto) [reply components/supervisor/pkg/supervisor/top.go:34]: At the moment the code in ws-daemon is not written to be used by other components. Factoring this out would be equivalent to starting our own cgroup library. Considering that the other options like using containerd's cgroup are awkward to use I still think this is the best way forward. This does not need to be done in this PR though.

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: Could you point me to relevant code please? 🙏 

There also can be slight difference, i.e. I subtract evict-able memory to give a user more usefull data, i.e. simlar to what GCP showing. I am not sure what ws-daemon does.

Comment (Furisto) [reply components/supervisor/pkg/supervisor/top.go:34]: CPU: 
v1; https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cpulimit/cfs.go
v2: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cpulimit/cfs_v2.go

Memory: 
v1: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cgroup/plugin_cachereclaim.go
v2: n.a.

IO:
v1: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cgroup/plugin_iolimit_v1.go
v2: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cgroup/plugin_iolimit_v2.go

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: > We should harmonise this with the way ws-daemon does it.

After looking at the code I don't think it can be harmonised without refactoring, ws daemon does not have some functions like getting used memory without evictable caches. Instead it compare page caches after trying to reclaim memory. It also has special cases for different state of workspaces which is not applicable inside a workspace.

I'm not sure refactoring will make such sense trying to create a reusable library right now. Code in ws-daemon is not really source of truth, i.e. there should not be a situation when you change one and need to change another.

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: I filed https://github.com/gitpod-io/gitpod/issues/10043 to look how to refactor it nicely later.

Comment (utam0k) [reply components/supervisor/pkg/supervisor/top.go:34]: @akosyakov Great PR. I have a question do you have a specific plan to support cgroup v2? As far as I know, In the case of cgroup v2, the supervisor may don't see the cgroup value of the pod.
https://github.com/gitpod-io/gitpod/blob/cf2c0aafa9d38973e3b599637e8c7256d8d3701b/components/ws-daemon/pkg/iws/iws.go#L915-L919
I recommend checking to be sure. You can build the gcp image with cgroup v2 by changing this line. And you can use a gcp image you build on workspace-preview.
https://github.com/gitpod-io/gitpod-packer-gcp-image/blob/ef5be5d24542142f39fc2aa6748761d2f2177124/setup.sh#L111-L112

Comment (utam0k) [reply components/supervisor/pkg/supervisor/top.go:34]: You can use `gitpod-k3s-202205162340` I built as a cgroup v2 version image.
on workspace
```console
root@gitpodio-templategolang-twmrukrpk05:/sys/fs/cgroup# cat /sys/fs/cgroup/memory.max 
max
root@gitpodio-templategolang-twmrukrpk05:/sys/fs/cgroup# cat /sys/fs/cgroup/memory.high
max
```
I make sure that the workspace's pod was limited
```console
    resources:
      limits:
        cpu: "2"
        memory: 4Gi
      requests:
        cpu: "1"
        memory: 2Gi
```

Comment (utam0k) [reply components/supervisor/pkg/supervisor/top.go:34]: So, I guess what I am saying is that in cgroup v2 the memory limit is not visible from the supervisor. This could be an issue with cgroup v2 as the JVM probably can't see it either.

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: @utam0k Is it always like that or it is how we implement it?

I asked already on other channels whether instead of relying on files we can actually ask container runtime endpoints to collect metrics and provide results via workspace kits somehow?

I wonder also why ws-daemon does not use container runtime endpoints but need to read from the disk?

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: Anyway we discussed with @iQQBot to land this PR as is today and then tomorrow start working on https://github.com/gitpod-io/gitpod/issues/10043 and support of cgroup v2 in supervisor. So I think we will need to research it as a part of it.

File: components/supervisor-api/status.proto
Comment (akosyakov) [reply components/supervisor-api/status.proto:76]: for now I will stick with internal cli, there are some uncertainties about implementation, so let's explore first

File: components/supervisor/pkg/supervisor/top.go
Comment (utam0k) [reply components/supervisor/pkg/supervisor/top.go:34]: > I wonder also why ws-daemon does not use container runtime endpoints but need to read from the disk?

Thanks for your feedback! Sorry, I couldn't get what `the disk` is.

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: > Thanks for your feedback! Sorry, I couldn't get what the disk is.

I mean container runtime has [such API](https://github.com/opencontainers/runc/blob/067aaf8548d78269dcb2c13b856775e27c410f9c/libcontainer/container.go#L94-L95) which can return usage/limit stats for cpu and memory: https://github.com/opencontainers/runc/blob/067aaf8548d78269dcb2c13b856775e27c410f9c/libcontainer/cgroups/stats.go#L158-L161

Supervisor cannot reach this endpoint, so we have to read cgroup from filesystem. I am asking whether ws-daemon can rely on it instead and provide this information to supervisor somehow?

Comment (Furisto) [reply components/supervisor/pkg/supervisor/top.go:34]: The API you have linked would mean to execute runc stat and then parse the output, this is not a service endpoint. We also need to write to the cgroup filesystem not only read from it and we should strive to do so in a runtime agnostic manner. We could ask this information from containerd but if we allow the option to use any CRI compatible runtime in the future (like CRI-O) this would require us to adapt the code again.

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: ok, sorry i don't really know it and though opencontainers runc is an abstraction over different runtimes and using this API oppositely will make things more stable.

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/top.go:34]: @csweichel Would it be alright to continue as suggested [here](https://github.com/gitpod-io/gitpod/pull/9796#discussion_r874502070)? Could you approve it please if so? 🙏

**Review by Furisto — APPROVED** [review]

**Comment by mustard-mh** [comment]
Code looks good, will test it with Prometheus

**Review by csweichel — APPROVED** [review]

**Review by mustard-mh — APPROVED** [review]
Tested good

|used|max|
|-|-|
|||

Another way to check xmx=4096m in thin client (related https://github.com/gitpod-io/gitpod/pull/9796#issuecomment-1125895833

|Img|
|-|
|image|

**Comment by mustard-mh** [comment]
Feel free to unhold it 👍 @akosyakov

**Review by utam0k — APPROVED** [review]

**Pull request merged**
